es/resource_elasticsearch_audit_config.go

```diff
@@ -0,0 +1,521 @@
+		Enabled:             m["enabled"].(bool),
+		InternalConfig:      m["internal_config"].(bool),
+		ExternalConfig:      m["external_config"].(bool),
+		WriteMetadataOnly:   m["write_metadata_only"].(bool),
+		ReadMetadataOnly:    m["read_metadata_only"].(bool),
+		WriteLogDiffs:       m["write_log_diffs"].(bool),
+		ReadWatchedFields:   expandReadWatchedFields(m["read_watched_field"].(*schema.Set).List()),
+		ReadIgnoreUsers:     expandStringList(m["read_ignore_users"].(*schema.Set).List()),
+		WriteWatchedIndices: expandStringList(m["write_watched_indices"].(*schema.Set).List()),
+		WriteIgnoreUsers:    expandStringList(m["write_ignore_users"].(*schema.Set).List()),
+	}
+}
+
+func expandReadWatchedFields(fields []interface{}) map[string][]string {
+	result := map[string][]string{}
+
+	for _, field := range fields {
+		key := field.(map[string]interface{})["index"].(string)
+		values := field.(map[string]interface{})["fields"].(*schema.Set).List()
+		result[key] = []string{}
+		for _, v := range values {
+			result[key] = append(result[key], v.(string))
+		}
+	}
+	return result
+}
+
+func resourceElasticsearchPutAuditConfig(d *schema.ResourceData, m interface{}) (*putResponse, error) {
+	response := new(putResponse)
+	auditConfig := config{
+		Enabled:    d.Get("enabled").(bool),
+		Audit:      expandAudit(d),
+		Compliance: expandCompliance(d),
+	}
+
+	auditConfigJSON, err := json.Marshal(auditConfig)
+	if err != nil {
+		return response, fmt.Errorf("body Error : %s", auditConfigJSON)
+	}
+
+	var body json.RawMessage
+	esClient, err := getClient(m.(*ProviderConf))
+	if err != nil {
+		return nil, err
+	}
+	switch client := esClient.(type) {
+	case *elastic7.Client:
+		var res *elastic7.Response
+		log.Printf("[INFO] put audit config: %+v", auditConfig)
+		res, err = client.PerformRequest(context.TODO(), elastic7.PerformRequestOptions{
+			Method:           "PUT",
+			Path:             "/_plugins/_security/api/audit/config",
+			Body:             string(auditConfigJSON),
+			RetryStatusCodes: []int{http.StatusInternalServerError},
+			Retrier: elastic7.NewBackoffRetrier(
+				elastic7.NewExponentialBackoff(100*time.Millisecond, 30*time.Second),
+			),
+		})
+		if err != nil {
+			e, ok := err.(*elastic7.Error)
+			if !ok {
+				log.Printf("[ERROR] expected error to be of type *elastic.Error")
+			} else {
+				log.Printf("[ERROR] error creating audit config: %v %v %v", res, res.Body, e)
+			}
+			return response, err
+		}
+
+		body = res.Body
+	default:
+		return response, errors.New("audit config resource not implemented prior to Elastic v7")
+	}
+
+	if err := json.Unmarshal(body, response); err != nil {
+		return response, fmt.Errorf("failed to unmarshal audit config body: %+v: %+v", err, body)
+	}
+
+	return response, nil
+}
+
+// Response used by the security plugin API (GET method)
+type getResponse struct {
+	Config config `json:"config"`
+}
+
+// Response sent by the security plugin API (PUT method)
+type putResponse struct {
```

Comment:
can we make these names more descriptive so it doesn't conflict with other responses, e.g. getAuditResponse, putAuditResponse, auditConfig, auditCompliance?